pyodide-build/pyodide_build/io.py

```diff
@@ -1,255 +1,224 @@
-    file_path
-        optional meta.yaml file path. Only used for more explicit error output,
-        when raise_errors = True.
-    """
-    yield from _check_config_keys(config)
-    yield from _check_config_types(config)
-    yield from _check_config_source(config)
-    yield from _check_config_build(config)
-    yield from _check_config_wheel_build(config)
-
-
-def check_package_config(
-    config: dict[str, Any],
-    file_path: Path | str | None = None,
-    raise_errors: bool = True,
-) -> list[str]:
-    errors_msg = list(check_package_config_generate_errors(config))
-
-    if errors_msg:
-        if file_path is None:
-            file_path = Path("meta.yaml")
-        if raise_errors:
+        if in_tree and from_url:
             raise ValueError(
-                f"{file_path} validation failed: \n  - " + "\n - ".join(errors_msg)
+                "Source section should not have both a 'url' and a 'path' key"
             )
-    return errors_msg
+        return values
 
+    @pydantic.root_validator
+    def _check_patches_extra(cls, values: dict[str, Any]) -> dict[str, Any]:
+        patches = values["patches"]
+        extras = values["extras"]
+        in_tree = values["path"] is not None
+        from_url = values["url"] is not None
 
-def parse_package_config(path: Path | str, *, check: bool = True) -> dict[str, Any]:
-    """Load a meta.yaml file
+        url_is_wheel = from_url and values["url"].endswith(".whl")
 
-    Parameters
-    ----------
-    path
-       path to the meta.yaml file
-    check
-       check the consistency of the config file
+        if in_tree and (patches or extras):
+            raise ValueError(
+                "If source is in tree, 'source/patches' and 'source/extras' keys "
+                "are not allowed"
+            )
 
-    Returns
-    -------
-    the loaded config as a Dict
-    """
-    # Import yaml here because pywasmcross needs to run in the built native
-    # Python, which won't have PyYAML
-    import yaml
+        if url_is_wheel and (patches or extras):
+            raise ValueError(
+                "If source is a wheel, 'source/patches' and 'source/extras' "
+                "keys are not allowed"
+            )
+
+        return values
+
+
+_BuildSpecExports = Literal["pyinit", "requested", "whole_archive"]
+
+
+class _BuildSpec(BaseModel):
+    exports: _BuildSpecExports | list[_BuildSpecExports] = "pyinit"
+    backend_flags: str = Field("", alias="backend-flags")
+    cflags: str = ""
+    cxxflags: str = ""
+    ldflags: str = ""
+    library: bool = False
+    sharedlibrary: bool = False
+    cross_script: str | None = Field(None, alias="cross-script")
+    script: str | None = None
+    post: str | None = None
+    unvendor_tests: bool = Field(True, alias="unvendor-tests")
+    cross_build_env: bool = Field(False, alias="cross-build-env")
+    cross_build_files: list[str] = Field([], alias="cross-build-files")
```

Comment:
There is a bit of a question about whether we keep these aliases or rename the meta.yaml fields to use `_`. If we rename them it should probably be a separate PR anyway.

Reply:
I have no strong opinion. It seems like YAML has no strong convention for this.